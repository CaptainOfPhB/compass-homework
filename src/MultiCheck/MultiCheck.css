.multi-check {
  min-width: 325px;
  display: inline-block;
  flex-direction: column;
}

.multi-check-label {
  padding: 5px 10px;
  background: rgba(128, 128, 128, .3);
}

.multi-check-options-outer {
  padding: 7px 10px;
  background: #f7f7f9;
}

.multi-check-options-inner {
  display: flex;
  background: white;
  border-radius: 3px;
  border: 1px solid #e6e6e6;
}

.multi-check-column {
  min-width: 150px;
  padding: 5px 10px;
}

.multi-check-column:not(:first-child) {
  margin-left: 5px;
}

.multi-check-option {
  display: flex;
  align-items: center;
  vertical-align: text-bottom;
}

.multi-check-option input {
  cursor: pointer;
}

.multi-check-option label {
  flex-grow: 1;
  cursor: pointer;
  margin-left: 5px;
}

.multi-check input[type="checkbox"] {
  width: 15px;
  height: 15px;
  accent-color: black;
}
