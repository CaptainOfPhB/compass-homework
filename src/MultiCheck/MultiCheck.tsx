import './MultiCheck.css';

import React, { useEffect, FC, useState, ChangeEvent } from 'react';

export type Option = {
  label: string,
  value: string
}

/**
 * Notice:
 * 1. There should be a special `Select All` option with checkbox to control all passing options
 * 2. If columns > 1, the options should be placed from top to bottom in each column
 *
 * @param {string} label - the label text of this component
 * @param {Option[]} options - options. Assume no duplicated labels or values
 * @param {string[]} values - If `undefined`, makes the component in uncontrolled mode with no options checked;
 *                            if not undefined, makes the component to controlled mode with corresponding options checked.
 *                            The values CAN be duplicated or NOT in the provided options
 * @param {number} columns - default value is 1. If it's bigger than all options count, make it same as the count of all options 
 * @param {Function} onChange - if not undefined, when checked options are changed, they should be passed to outside;
 *                              if undefined, the options can still be selected, but won't notify the outside
 * @param {Function} onRender - will be called if current component rendered. Determine the balance between a reasonable
 *                              render count and readable code
 */
export type Props = {
  label?: string,
  options: Option[],
  columns?: number,
  values?: string[]
  onChange?: (options: Option[]) => void,
  onRender: () => void
}

export const MultiCheck: FC<Props> = (props) => {
  const {onRender} = props;

  {
    // NOTE Don't modify the code block, it can be considered as a performance hint,
    //      you need to find a way to avoid triggering it infinitely
    console.log('### > MultiCheck');
    useEffect(() => {
      onRender()
    })
  }

  const [layout, setLayout] = useState<number[]>([]);
  const [selectAllChecked, setSelectAllChecked] = useState<boolean>(false);
  const [selectedValues, setSelectedValues] = useState<string[]>(props.values || []);

  function onSelectAllChange(event: ChangeEvent<HTMLInputElement>) {
    const checked = event.target.checked;
    const finialSelectedValues = checked ? props.options.map(option => option.value) : [];
    setSelectAllChecked(checked);
    setSelectedValues(finialSelectedValues);
    outsideOnChange(finialSelectedValues);
  }

  function onCheckboxChange(option: Option, event: ChangeEvent<HTMLInputElement>) {
    const checked = event.target.checked;
    const finalSelectedValues = checked
      ? selectedValues.concat(option.value)
      : selectedValues.filter(value => value !== option.value);
    setSelectedValues(finalSelectedValues);
    shouldSelectAllChecked(finalSelectedValues);
    outsideOnChange(finalSelectedValues);
  }

  function shouldSelectAllChecked(finalSelectedValues: string[]) {
    const allOptionsChecked = (props.options || []).every(option => finalSelectedValues.includes(option.value));
    setSelectAllChecked(allOptionsChecked);
  }

  function outsideOnChange(finalSelectedValues: string[]) {
    if (props.onChange) {
      props.onChange(props.options.filter(option => finalSelectedValues.includes(option.value)));
    }
  }

  function renderOptions() {
    const finalOptions = [...props.options];
    return layout.map((count, column) => {
      const options = finalOptions.splice(0, count);
      return (
        <div key={`column-${column}`} className='multi-check-column'>
          {/* The "Select All" option */}
          {column === 0
            ? (
              <div className='multi-check-option multi-check-select-all'>
                <input
                  type='checkbox'
                  id='multi-check-select-all'
                  name='multi-check-select-all'
                  checked={selectAllChecked}
                  onChange={onSelectAllChange}
                />
                <label htmlFor='multi-check-select-all'>Select&nbsp;All</label>
              </div>
            ) : null
          }
          {options.map(option => {
            const checked = selectedValues.includes(option.value);
            return (
              <div key={option.value} className='multi-check-option'>
                <input
                  type='checkbox'
                  id={option.label}
                  checked={checked}
                  name={option.label}
                  onChange={e => onCheckboxChange(option, e)}
                />
                <label htmlFor={option.label}>{option.label}</label>
              </div>
            );
          })}
        </div>
      );
    });
  }

  useEffect(function setValuesIfChanged() {
    setSelectedValues(props.values || []);
    shouldSelectAllChecked(props.values || []);
  }, [props.values]);

  useEffect(function geneLayout() {
    // don't forget the extra "Select All" option
    const optionsTotalCount = props.options.length + 1;
    // in case of negative or float number
    const columnCount = props.columns && props.columns > 0 ? parseInt(props.columns.toString()) : 1;
    const intactRows = parseInt((optionsTotalCount / columnCount).toString());
    const restOptionsCount = optionsTotalCount % columnCount;
    // each item is the number of the option that should be rendered for each column
    let layout = Array(columnCount).fill(intactRows);
    // put rest options in order into each column
    for (let i = 0; i < restOptionsCount; i++) {
      layout[i] += 1;
    }
    // then, remove the count for the "Select All" option
    layout.splice(0, 1, layout[0] - 1);
    setLayout(layout);
  }, [props.columns, props.options]);

  return (
    <div className='multi-check'>
      {props.label
        ? <div className='multi-check-label'>{props.label}</div>
        : null
      }
      {props.options && props.options.length
        ? (
          <div className='multi-check-options-outer'>
            <div className='multi-check-options-inner'>
              {renderOptions()}
            </div>
          </div>
        )
        : 'No options provided'
      }
    </div>
  );
};
