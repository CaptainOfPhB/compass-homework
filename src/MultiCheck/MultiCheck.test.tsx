import React from 'react';
import { MultiCheck } from './MultiCheck';
import { fireEvent, render, screen, cleanup } from '@testing-library/react';

describe('MultiCheck', () => {
  it('renders the label if label provided', () => {
    const onRender = jest.fn();
    const label = 'test label text';
    render(<MultiCheck label={label} options={[]} onRender={onRender} />);
    expect(screen.getByText(label)).toBeInTheDocument();
  });

  it('onRender should be called', () => {
    const onRender = jest.fn();
    render(<MultiCheck options={[]} onRender={onRender} />);
    expect(onRender).toBeCalledTimes(2);
  });

  it('render options if provided', () => {
    const onRender = jest.fn();
    const options = [{ label: 'label', value: 'value' }];
    render(<MultiCheck options={options} onRender={onRender} />);
    expect(screen.getByText('label')).toBeInTheDocument();
  });

  it('render default placeholder if no options provided', () => {
    const onRender = jest.fn();
    render(<MultiCheck options={[]} onRender={onRender} />);
    expect(screen.getByText('No options provided')).toBeInTheDocument();
  });

  it('onChange should be called', () => {
    const onRender = jest.fn();
    const onChange = jest.fn();
    const options = [{ label: 'label', value: 'value' }];
    render(<MultiCheck options={options} onChange={onChange} onRender={onRender} />);
    fireEvent.click(screen.getByText('label'));
    expect(onChange).toBeCalledWith(options);
    fireEvent.click(screen.getByText('label'));
    expect(onChange).toBeCalledWith([]);
  });

  it('all options should be checked', () => {
    const onRender = jest.fn();
    const options = [
      { label: 'label1', value: 'value1' },
      { label: 'label2', value: 'value2' },
    ];
    const values = options.map(option => option.value);
    render(<MultiCheck values={values} options={options} onRender={onRender} />);
    const allCheckboxes = screen.getAllByRole('multi-check-option-checkbox');
    const allChecked = allCheckboxes.every((checkbox: HTMLElement) => (checkbox as HTMLInputElement).checked);
    const selectAllCheckbox = screen.getByRole('multi-check-select-all');
    const selectAllChecked = (selectAllCheckbox as HTMLInputElement).checked;
    expect(allChecked).toBe(true);
    expect(selectAllChecked).toBe(true);
  });

  it('partial options should be checked', () => {
    const onRender = jest.fn();
    const options = [
      { label: 'label1', value: 'value1' },
      { label: 'label2', value: 'value2' },
    ];
    const values = [options[0].value];
    render(<MultiCheck values={values} options={options} onRender={onRender} />);
    const targetCheckbox = screen.getByTestId(options[0].label) as HTMLInputElement;
    const selectAllCheckbox = screen.getByRole('multi-check-select-all') as HTMLInputElement;
    expect(targetCheckbox.checked).toBe(true);
    expect(selectAllCheckbox.checked).toBe(false);
  });

  it('select all should work', async () => {
    const onRender = jest.fn();
    const options = [
      { label: 'label1', value: 'value1' },
      { label: 'label2', value: 'value2' },
    ];
    const getAllCheckboxes = () => screen.getAllByRole('multi-check-option-checkbox');
    const isChecked = (checkbox: HTMLElement) => (checkbox as HTMLInputElement).checked;
    render(<MultiCheck options={options} onRender={onRender} />);
    const selectAllCheckbox = screen.getByRole('multi-check-select-all');
    fireEvent.click(selectAllCheckbox);
    const allChecked = getAllCheckboxes().every(isChecked);
    fireEvent.click(selectAllCheckbox);
    const anyChecked = getAllCheckboxes().some(isChecked);
    expect(allChecked).toBe(true);
    expect(anyChecked).toBe(false);
  });

  it('minimum columns is 1', () => {
    const onRender = jest.fn();
    const options = [
      { label: 'label1', value: 'value1' },
      { label: 'label2', value: 'value2' },
      { label: 'label3', value: 'value3' },
      { label: 'label4', value: 'value4' },
      { label: 'label5', value: 'value5' },
      { label: 'label6', value: 'value6' },
    ];
    render(<MultiCheck options={options} onRender={onRender} />);
    const columns1 = screen.getAllByTestId('multi-check-column');
    cleanup();
    render(<MultiCheck columns={-1} options={options} onRender={onRender} />);
    const columns2 = screen.getAllByTestId('multi-check-column');
    cleanup();
    render(<MultiCheck columns={1.1} options={options} onRender={onRender} />);
    const columns3 = screen.getAllByTestId('multi-check-column');
    cleanup();
    expect(columns1.length).toBe(1);
    expect(columns2.length).toBe(1);
    expect(columns3.length).toBe(1);
  });

  it('layout order is up to bottom and left to right', () => {
    const onRender = jest.fn();
    const options = [
      { label: 'label1', value: 'value1' },
      { label: 'label2', value: 'value2' },
      { label: 'label3', value: 'value3' },
      { label: 'label4', value: 'value4' },
      { label: 'label5', value: 'value5' },
      { label: 'label6', value: 'value6' },
    ];
    render(<MultiCheck columns={2} options={options} onRender={onRender} />);
    const columns1 = screen.getAllByTestId('multi-check-column');
    cleanup();
    render(<MultiCheck columns={3} options={options} onRender={onRender} />);
    const columns2 = screen.getAllByTestId('multi-check-column');
    cleanup();
    render(<MultiCheck columns={4} options={options} onRender={onRender} />);
    const columns3 = screen.getAllByTestId('multi-check-column');
    cleanup();
    expect(columns1).toMatchSnapshot();
    expect(columns2).toMatchSnapshot();
    expect(columns3).toMatchSnapshot();
    expect(columns1.length).toBe(2);
    expect(columns2.length).toBe(3);
    expect(columns3.length).toBe(4);
  });
});
