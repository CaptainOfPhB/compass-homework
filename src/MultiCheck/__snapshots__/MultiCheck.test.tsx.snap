// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiCheck layout order is up to bottom and left to right 1`] = `
Array [
  <div
    class="multi-check-column"
    data-testid="multi-check-column"
  >
    <div
      class="multi-check-option multi-check-select-all"
    >
      <input
        id="multi-check-select-all"
        name="multi-check-select-all"
        role="multi-check-select-all"
        type="checkbox"
      />
      <label
        for="multi-check-select-all"
      >
        Select All
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label1"
        id="label1"
        name="label1"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label1"
      >
        label1
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label2"
        id="label2"
        name="label2"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label2"
      >
        label2
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label3"
        id="label3"
        name="label3"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label3"
      >
        label3
      </label>
    </div>
  </div>,
  <div
    class="multi-check-column"
    data-testid="multi-check-column"
  >
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label4"
        id="label4"
        name="label4"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label4"
      >
        label4
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label5"
        id="label5"
        name="label5"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label5"
      >
        label5
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label6"
        id="label6"
        name="label6"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label6"
      >
        label6
      </label>
    </div>
  </div>,
]
`;

exports[`MultiCheck layout order is up to bottom and left to right 2`] = `
Array [
  <div
    class="multi-check-column"
    data-testid="multi-check-column"
  >
    <div
      class="multi-check-option multi-check-select-all"
    >
      <input
        id="multi-check-select-all"
        name="multi-check-select-all"
        role="multi-check-select-all"
        type="checkbox"
      />
      <label
        for="multi-check-select-all"
      >
        Select All
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label1"
        id="label1"
        name="label1"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label1"
      >
        label1
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label2"
        id="label2"
        name="label2"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label2"
      >
        label2
      </label>
    </div>
  </div>,
  <div
    class="multi-check-column"
    data-testid="multi-check-column"
  >
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label3"
        id="label3"
        name="label3"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label3"
      >
        label3
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label4"
        id="label4"
        name="label4"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label4"
      >
        label4
      </label>
    </div>
  </div>,
  <div
    class="multi-check-column"
    data-testid="multi-check-column"
  >
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label5"
        id="label5"
        name="label5"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label5"
      >
        label5
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label6"
        id="label6"
        name="label6"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label6"
      >
        label6
      </label>
    </div>
  </div>,
]
`;

exports[`MultiCheck layout order is up to bottom and left to right 3`] = `
Array [
  <div
    class="multi-check-column"
    data-testid="multi-check-column"
  >
    <div
      class="multi-check-option multi-check-select-all"
    >
      <input
        id="multi-check-select-all"
        name="multi-check-select-all"
        role="multi-check-select-all"
        type="checkbox"
      />
      <label
        for="multi-check-select-all"
      >
        Select All
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label1"
        id="label1"
        name="label1"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label1"
      >
        label1
      </label>
    </div>
  </div>,
  <div
    class="multi-check-column"
    data-testid="multi-check-column"
  >
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label2"
        id="label2"
        name="label2"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label2"
      >
        label2
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label3"
        id="label3"
        name="label3"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label3"
      >
        label3
      </label>
    </div>
  </div>,
  <div
    class="multi-check-column"
    data-testid="multi-check-column"
  >
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label4"
        id="label4"
        name="label4"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label4"
      >
        label4
      </label>
    </div>
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label5"
        id="label5"
        name="label5"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label5"
      >
        label5
      </label>
    </div>
  </div>,
  <div
    class="multi-check-column"
    data-testid="multi-check-column"
  >
    <div
      class="multi-check-option"
    >
      <input
        data-testid="label6"
        id="label6"
        name="label6"
        role="multi-check-option-checkbox"
        type="checkbox"
      />
      <label
        data-testid="multi-check-option-label"
        for="label6"
      >
        label6
      </label>
    </div>
  </div>,
]
`;
